xcm/xcm-executor/src/traits/on_response.rs

```diff
@@ -94,3 +102,73 @@ impl VersionChangeNotifier for () {
 		false
 	}
 }
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum FinishedQuery<BlockNumber> {
+	Response { response: Response, at: BlockNumber },
+	VersionNotification { origin: MultiLocation, is_active: bool },
+}
+
+/// The possible state of an XCM query response.
+#[derive(Debug, PartialEq, Eq)]
+pub enum QueryResponseStatus<BlockNumber> {
+	/// The response has arrived, and includes the inner Response and the block number it arrived at.
+	Finished(FinishedQuery<BlockNumber>),
+	/// The response has not yet arrived, the XCM might still be executing or the response might be in transit.
+	Pending,
+	/// No response with the given `QueryId` was found, or the response was already queried and removed from local storage.
+	NotFound,
+	/// Got an unexpected XCM version.
+	UnexpectedVersion,
+}
+
+/// Provides methods to expect responses from XCMs and query their status.
+pub trait XcmQueryHandler {
+	type QueryId: From<u64>
+		+ FullCodec
+		+ MaxEncodedLen
+		+ TypeInfo
+		+ Clone
+		+ Eq
+		+ PartialEq
+		+ Debug
+		+ Copy;
+	type BlockNumber;
+	type Error;
+	type UniversalLocation: Get<InteriorMultiLocation>;
+
+	/// Attempt to create a new query ID and register it as a query that is yet to respond.
+	fn new_query(
+		responder: impl Into<MultiLocation>,
+		timeout: Self::BlockNumber,
+		match_querier: impl Into<MultiLocation>,
+	) -> QueryId;
+
+	/// Consume `message` and return another which is equivalent to it except that it reports
+	/// back the outcome.
+	///
+	/// - `message`: The message whose outcome should be reported.
+	/// - `responder`: The origin from which a response should be expected.
+	/// - `timeout`: The block number after which it is permissible to return `NotFound` from `take_response`.
+	///
+	/// `report_outcome` may return an error if the `responder` is not invertible.
+	///
+	/// It is assumed that the querier of the response will be `Here`.
+	/// The response can be queried with `take_response`.
+	fn report_outcome(
+		message: &mut Xcm<()>,
+		responder: impl Into<MultiLocation>,
+		timeout: Self::BlockNumber,
+	) -> result::Result<Self::QueryId, Self::Error>;
+
+	/// Makes sure to expect a response with the given id
+	/// Used for bencharks.
+	fn expect_response(id: Self::QueryId);
+
+	/// Attempt to remove and return the response of query with ID `query_id`.
+	///
+	/// Returns `Finished` if the response is available and includes inner response.
+	/// Returns `Pending` if the response is not yet available.
+	/// Returns `NotFound` if the response is not available and won't ever be.
+	fn take_response(id: Self::QueryId) -> QueryResponseStatus<Self::BlockNumber>;
```

Comment:
this is old name from the pallet, but its not really take_response anymore

Reply:
Don't see anything wrong with it being in the same trait as the response-handler code(?), but I have renamed the trait to make the breadth of its responsibility clearer.

Reply:
in general we do expect it to elicit an eventual removal, no?